Optimze and new first layer calibration

3d-gussner opened the pull request:
- Reduced buffer and flash size by
  - changing to relative mode during the first layer calibration pattern
  - "compressing" the Gcode line by removing spaces 
- added NEW_FIRST_LAYER_CAL to perform 1st layer cal from front to rear as requested in https://github.com/prusa3d/Prusa-Firmware/issues/4572

By default the previous known first layer calibration pattern is used. You need to un-comment `//#define NEW_FIRST_LAYER_CAL` in the variant file to use the new pattern from front to rear.

The new first layer calibration pattern from front to rear with square ends in the middle of the bed.
- The calibration of the first layer at the center is most important
- The left, right, front rear sides can manually corrected via LCD -> Calibration -> Bed level correct menu
- The first layer calibration pattern continues directly after the purge to keep the pressure stable in the nozzle

![New_first_layer_cal](https://github.com/user-attachments/assets/3beb9b61-6aa4-49c8-9c7e-3236044eb8d1)

3d-gussner committed:
New first layer calibration
- changed to relative mode to save buffer and code size
- added NEW_FIRST_LAYER_CAL to perform 1st layer cal from front to back

3d-gussner commented:
First build

 All values in bytes. Δ Delta to base
| Target| ΔFlash | ΔSRAM | Used Flash | Used SRAM | Free Flash | Free SRAM |
| ------| ------ | ----- | -----------| --------- | ---------- | --------- |
| `MK3S_MULTILANG` | -668 | 0 | 247634 | 5654 | 6318 | 2538 |
| `MK3_MULTILANG` | -662 | 0 | 247010 | 5663 | 6942 | 2529 |

 That is more than expected. Running a rebase.

Still more than I expected, but I compiled mostly the new first layer cal pattern version
All values in bytes. Δ Delta to base

| Target| ΔFlash | ΔSRAM | Used Flash | Used SRAM | Free Flash | Free SRAM |
| ------| ------ | ----- | -----------| --------- | ---------- | --------- |
| `MK3S_MULTILANG` | -412 | 0 | 247890 | 5654 | 6062 | 2538 |
| `MK3_MULTILANG` | -412 | 0 | 247260 | 5663 | 6692 | 2529 |

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -210,41 +224,37 @@ void lay1cal_meander(float layer_height, float extrusion_width)
     lay1cal_common_enqueue_loop(cmd_cal_finish, (sizeof(cmd_cal_finish)/sizeof(cmd_cal_finish[0])));
 
     if (mmu_enabled) enquecommand_P(MSG_M702); //unload from nozzle
+    disable_heater();
```
Note that MMU M702 needs the heater to be on to complete the unload. The heaters are already turned off above with  `M140 S0` (See `cmd_cal_finish_2`) and `M104 S0` (See `cmd_cal_finish_3`)

We can then also avoid including `#include "temperature.h"`

Now that I look at the code, the `M702` g-code should actually come before `M104 S0`

3d-gussner replied:
Changing to `disable_heater();` saves few additional bytes.

leptun replied:
I agree with @gudnimg. We should not disable the heaters before the M702. Moreover, directly calling disable_heater bypasses the cmdqueue, so stuff ends up executing out of order. Please switch back to M104 and M140, but after the M702 stuff. I know it will add back some bytes, but I would rather have that than mix and match cmdqueue and non-cmdqueue actions in the automation and cause issues with the M702 unload.

~~Also grrrrrr the MK3SMMU3 prusaslicer profiles also do this out of order :/~~ I can't read. Ignore that :>

leptun replied:
I'd like to see the changes from #4813 in this PR to address this issue.

gudnimg replied:
I’ll prepare a PR on @3d-gussner ‘s fork soon

leptun replied:
Merged. Should be resolved now. @3d-gussner please check this in the sim.

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -96,65 +107,61 @@ bool lay1cal_load_filament(uint8_t filament)
-    static const char cmd_intro_mmu_9[] PROGMEM = "G1 Y-2 F1000";
-    static const char cmd_intro_mmu_10[] PROGMEM = "G1 X55 E25 F1400";
-    static const char cmd_intro_mmu_11[] PROGMEM = "G1 Z0.2 F1000";
-    static const char cmd_intro_mmu_12[] PROGMEM = "G1 X5 E4 F1000";
```
Is `cmd_intro_mmu_12` removal intentional? Also `cmd_intro_mmu_10` is changed, why?

3d-gussner replied:
Yes both with and without MMU purge line end at ~same position to continue with the new first layer cal pattern directly after the purge to keep the nozzle pressure up.

leptun replied:
I'm not sure if reducing the MMU purge line will have any unintended side effects. It could be a problem is switching between material types, but I suppose the user should pay attention when doing that anyway. Color bleed isn't really of significance on the first layer calibration, so I think this should be ok. Still I propose testing play around with this with material changes with the reduced purge.

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -96,65 +107,61 @@ bool lay1cal_load_filament(uint8_t filament)
-        static const char fmt1[] PROGMEM = "G1 X%d E%-.3f F1000";
-        enquecommandf_P(fmt1, 60, count_e(layer_height, extrusion_width * 4.f, 60));
-        enquecommandf_P(fmt1, 100, count_e(layer_height, extrusion_width * 8.f, 40));
+        enquecommandf_P(extrude_fmt_X, (float)60, count_e(layer_height, extrusion_width * 4.f, 60));
+        enquecommandf_P(extrude_fmt_X, (float)202.5, count_e(layer_height, extrusion_width * 8.f, 142.5));
```
This new format string removed the feedrate settings, previously it was 1000 mm/min or 16.67 mm/s

This previous g-code is setting 1000 mm/min so in this case it's OK. But if that string is changed, then this g-code will also change with it.

3d-gussner replied:
If no feedrate is provided the last set feedrate is used. Saves few bytes and the printed lines look the same as before.

leptun replied:
I agree with @gudnimg that we should probably set the feedrate to something known here once before the non-mmu purge line. No clue if something before it sets the feedrate to 1000. And yea, not specifying the F every command is fine, but we should still set the F before these two lines just for consistency at the expense of ~30B of flash.
From what I see quickly looking through the code, nothing before this sets the feedrate, so these two lines would run at whatever feedrate was set before this, which could be reasonable (default feedrate) or something ridiculously high if something else executed before it. Please enqueue a `G1F<something>` before enqueueing any other commands in the non-mmu intro.

leptun replied:
Pushed, please double check.

gudnimg replied:
Another minor nitpick.

You can change:

`(float)60` to `60.f`
`(float)202.5` to` 202.5f`

Slightly shorter code.

leptun replied:
@gudnimg fixed nitpick

leptun commented on Firmware/first_lay_cal.cpp:
```diff
@@ -96,65 +107,61 @@ bool lay1cal_load_filament(uint8_t filament)
             cmd_pre_meander_5,
+#endif //NEW_FIRST_LAYER_CAL
             cmd_pre_meander_6,
-            cmd_pre_meander_7,
```
The travel move will be at F7200, but I agree with this change. I always felt the travel is too slow for the first layer cal, too much oozing can happen.

leptun commented on Firmware/first_lay_cal.cpp:
```diff
@@ -163,42 +170,49 @@ void lay1cal_before_meander()
+    enquecommandf_P(extrude_fmt_Y, short_length, count_e(layer_height, extrusion_width, short_length));
+#endif //_NEW_FIRST_LAYER_CAL
+
+    enquecommand_P(PSTR("G91"));
```
maybe extract this to messages.h/cpp? - done

leptun replied:
Also, shouldn't this line be above the first meander move for the new calibration? This feels slightly misplaced.

leptun replied:
Fixed @3d-gussner

leptun commented on Firmware/first_lay_cal.cpp:
```diff
@@ -163,42 +170,49 @@ void lay1cal_before_meander()
-    enquecommandf_P(extrude_fmt, 100, 155, count_e(layer_height, extrusion_width * 2.f, 25));
-    enquecommandf_P(extrude_fmt, 200, 155, count_e(layer_height, extrusion_width, 100));
-    enquecommandf_P(extrude_fmt, 200, 135, count_e(layer_height, extrusion_width, 20));
+    enquecommand_P(PSTR("G1F1080"));
```
interesting choice for F. Could we maybe have a global G1F1000 that we can reuse here? Saves a few more bytes of flash and addresses the issue I mentioned earlier.

leptun replied:
Changed to F1080 in multiple places and reused the same command in progmem

leptun commented on Firmware/first_lay_cal.cpp:
```diff
@@ -210,41 +224,37 @@ void lay1cal_meander(float layer_height, float extrusion_width)
 //! @param i iteration
 void lay1cal_square(uint8_t step, float layer_height, float extrusion_width)
```
step parameter is now useless. Please remove it and fix the for loop to 4 cycles. Update comment above also and save some more flash by reducing parameter count :>

leptun replied:
done, saved 74B of flash

gudnimg committed:
Fix an issue where MMU unloads filament while its cooling down

Change in memory:
Flash: -6 bytes
SRAM: 0 bytes

gudnimg committed:
Unload the filament at parked position

We don't want to unload while the nozzle is still smudged against the bed.

leptun committed:
remove unused step parameter in lay1cal sqaure

leptun committed:
Ensure that a correct feedrate is used at all times

leptun committed:
Extract G91 to messages.h/cpp

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -857,16 +857,16 @@ void lcd_commands()
                 break;
             case 3:
-                lay1cal_square(3, layer_height, extrusion_width);
+                lay1cal_square(layer_height, extrusion_width);
```
Surely this can be optimised now :)? `layer_height` and `extrusion_width` are defined as constants.

Prehaps:


```c
case 6:
case 5:
case 4:
case 3:
    lay1cal_square(layer_height, extrusion_width);
    break;
```

leptun replied:
I think the plan is to have it account for the nozzle size. Layer height is indeed always 0.2mm but the extrusion width is changed based on the hardware setup.

gudnimg replied:
Indeed. But we can assume the nozzle diameter wont change during first layer calibration.

3d-gussner replied:
No it doesn't change.

gudnimg replied:
Made a PR: https://github.com/3d-gussner/Prusa-Firmware/pull/97

leptun replied:
merged

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -37,12 +37,30 @@ static constexpr float spacing(float layer_height, float extrusion_width, float
+static const char extrude_fmt_Y[] PROGMEM = "G1Y%gE%g";
+static const char zero_extrusion[] PROGMEM = "G92E0";
+static const char feedrate_F1080[] PROGMEM = "G1F1080";
+#ifndef NEW_FIRST_LAYER_CAL
+int8_t invert = 1;
+const float short_length = 20;
+#else
+int8_t invert = -1;
```
Is there any point having the same variable with the same value here regardless if `NEW_FIRST_LAYER_CAL` is defined or not.

Can `int8_t invert = 1;` be moved out of the `ifdef`? It doesn't save any memory, but its shorter code.

3d-gussner replied:
The values are `-1`  vs `1` to invert the direction

gudnimg replied:
Sorry I thought I saw both as -1 😅 false alarm.

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -163,88 +178,93 @@ void lay1cal_before_meander()
+
+    enquecommand_P(MSG_G91); //enable relative XYZ
+#ifndef NEW_FIRST_LAYER_CAL
+    enquecommandf_P(extrude_fmt_X, (float)25*invert, count_e(layer_height, extrusion_width * 4.f, 25));
+    enquecommandf_P(extrude_fmt_X, (float)25*invert, count_e(layer_height, extrusion_width * 2.f, 25));
+    enquecommandf_P(extrude_fmt_X, (float)100*invert, count_e(layer_height, extrusion_width, 100));
+    enquecommandf_P(extrude_fmt_Y, (float)-20*invert, count_e(layer_height, extrusion_width, 20));
```
I think dropping the float cast would be nice:

`(float)25*invert` can be change to `invert * 25.f`

Not critical, just nitpicking :)

leptun replied:
fixed

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -163,88 +178,93 @@ void lay1cal_before_meander()
+        enquecommandf_P(extrude_fmt_Y, -short_length*invert, short_extrusion);
     }
+#ifdef NEW_FIRST_LAYER_CAL
+    enquecommandf_P(extrude_fmt_X, -(long_length/2-square_width/2), long_extrusion); //~Middle of bed X125
```
Consider using multiplication instead:

`long_length/2` -> `long_length * 0.5f`

It may be the compiler is smart enough to do this change automatically, but lets be explicit. Multiplication is more efficient than division.

leptun replied:
Woops, commented on the wrong thing.

This line is overextruded since a shorter length is used with the same E count. Needs adjusting

![image](https://github.com/user-attachments/assets/19e5a2a5-9399-4342-a6a7-9bd7f8cf67d4)

3d-gussner replied:
@leptun Good point. the extrusion needs to be 0.5 of the long_length

leptun replied:
Are you sure? Feels to me like it would be better but still not 100% correct as that would still slightly overextrude. Remember that the line is (half of long_length - half of the square width). So just setting it to 0.5x will not account for the square size

leptun replied:
> Consider using multiplication instead:
> 
> `long_length/2` -> `long_length * 0.5f`
> 
> It may be the compiler is smart enough to do this change automatically, but lets be explicit. Multiplication is more efficient than division.

All divisions are done inside constexpr or on constant compile time computed variables, so I am not worried about these.

leptun replied:
@3d-gussner fixed overextrusion with new pattern. Please check

gudnimg replied:
Reminder for myself: check if that division is done at compile time. If not, is the division done once or twice.

leptun replied:
It was. I extracted that to a constexpr and no size change happened. After that I also switched to 0.5f*. Changed in https://github.com/prusa3d/Prusa-Firmware/pull/4811/commits/7a47bf30be79c6b9e9c042dbd6384eccad28a592

leptun commented on Firmware/first_lay_cal.cpp:
```diff
@@ -163,88 +178,93 @@ void lay1cal_before_meander()
+    enquecommandf_P(extrude_fmt_X, (float)100*invert, count_e(layer_height, extrusion_width, 100));
+    enquecommandf_P(extrude_fmt_Y, (float)-20*invert, count_e(layer_height, extrusion_width, 20));
+#else
+    enquecommandf_P(extrude_fmt_X, long_length*invert, count_e(layer_height, extrusion_width, long_length));
```
this line is overextruded since a shorter length is used with the same E count. Needs adjusting

![image](https://github.com/user-attachments/assets/19e5a2a5-9399-4342-a6a7-9bd7f8cf67d4)

gudnimg committed:
optimisation: combine a few functions calls into one

Change in memory:
Flash: -34 bytes
SRAM: 0 bytes

gudnimg commented on Firmware/first_lay_cal.cpp:
```diff
@@ -37,12 +37,30 @@ static constexpr float spacing(float layer_height, float extrusion_width, float
 
-static const char extrude_fmt[] PROGMEM = "G1 X%d Y%d E%-.5f";
-static const char zero_extrusion[] PROGMEM = "G92 E0";
+static const char extrude_fmt_X[] PROGMEM = "G1X%gE%g";
```
Does G1 support scientific notation? `g` means `Use the shortest representation: %e or %f`. Is something like 3.9265e+2 parsed correctly?

leptun replied:
nope. If we end up scientific notation then this will fail for sure. Maybe just a simple %f might be better suited. That doesn't pad the end with zeros, no?

3d-gussner replied:
I don't think that the avr will produce any `e+2` and also never seen anything longer than 7 chars `"G1Y23.6111E0.74031"`

leptun replied:
Can you try to use %f and see if it generates the same commands? I think it will and will definitely be safer to use as it should always work.

leptun committed:
Get rid of float casts

leptun committed:
Minor nitpicks

leptun committed:
Ensure all meander moves are done using relative mode

Without this the first vertical move of the new meander would be slightly longer than expected due to using absolute moves from a negative Y starting position

leptun committed:
Optimize meander moves

leptun committed:
Fix end of meander overextrusion with new pattern

3d-gussner committed:
Switch back from `%g` to `.4f`
Set short length so the square is printed in the middle of the bed
Both purge lines with and without MMU end at X=202.5 to get the square in the middle of the bed

gudnimg reviewed (approved):
Approving, I'd like to see @leptun approve as well. Or at least 2nd approval :)

leptun reviewed (approved):
Tested both calibration versions in MK404 with fancy extrusion and colored extrusion based on width. Looks good to me. Only tested with 0.4mm nozzle

3d-gussner commented:
@gudnimg @leptun Thanks for the feedback and review.

jedisct1 commented:
Been running my MK3S+ with `NEW_FIRST_LAYER_CAL` for a while, and it works perfectly.
